subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/services/CacheInstrumentationDataBuildService.java

```diff
@@ -103,11 +98,13 @@ public ResolutionScope newResolutionScope(long contextId) {
             public void setTypeHierarchyAnalysisResult(FileCollection analysisResult) {
                 FileCollection typeHierarchyAnalysisResult = analysisResult.filter(InstrumentationTransformUtils::isTypeHierarchyAnalysisFile);
                 resolutionData.getTypeHierarchyAnalysisResult().setFrom(typeHierarchyAnalysisResult);
+                resolutionData.getTypeHierarchyAnalysisResult().finalizeValueOnRead();
```

Comment:
💭 I added finalization because otherwise we run InstumentationAnalysisTransform every time we query `getTypeHierarchyAnalysisResult()`. Even if it's up-to-date we can avoid that.